<h1>Содержание<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Исследование-базы-данных-книжного-приложения" data-toc-modified-id="Исследование-базы-данных-книжного-приложения-0"><span class="toc-item-num">0&nbsp;&nbsp;</span>Исследование базы данных книжного приложения</a></span></li><li><span><a href="#Описание-проекта" data-toc-modified-id="Описание-проекта-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Описание проекта</a></span><ul class="toc-item"><li><span><a href="#Описание-данных" data-toc-modified-id="Описание-данных-1.1"><span class="toc-item-num">1.1&nbsp;&nbsp;</span>Описание данных</a></span></li></ul></li><li><span><a href="#Ознакомление-с-данными" data-toc-modified-id="Ознакомление-с-данными-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Ознакомление с данными</a></span><ul class="toc-item"><li><span><a href="#Таблица-book" data-toc-modified-id="Таблица-book-2.1"><span class="toc-item-num">2.1&nbsp;&nbsp;</span>Таблица book</a></span></li><li><span><a href="#Таблица-authors" data-toc-modified-id="Таблица-authors-2.2"><span class="toc-item-num">2.2&nbsp;&nbsp;</span>Таблица authors</a></span></li><li><span><a href="#Таблица-publisher" data-toc-modified-id="Таблица-publisher-2.3"><span class="toc-item-num">2.3&nbsp;&nbsp;</span>Таблица publisher</a></span></li><li><span><a href="#Таблица-ratings" data-toc-modified-id="Таблица-ratings-2.4"><span class="toc-item-num">2.4&nbsp;&nbsp;</span>Таблица ratings</a></span></li><li><span><a href="#Таблица-reviews" data-toc-modified-id="Таблица-reviews-2.5"><span class="toc-item-num">2.5&nbsp;&nbsp;</span>Таблица reviews</a></span></li></ul></li><li><span><a href="#Задания" data-toc-modified-id="Задания-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Задания</a></span><ul class="toc-item"><li><span><a href="#Сколько-книг-вышло-после-1-января-2000-года" data-toc-modified-id="Сколько-книг-вышло-после-1-января-2000-года-3.1"><span class="toc-item-num">3.1&nbsp;&nbsp;</span>Сколько книг вышло после 1 января 2000 года</a></span></li><li><span><a href="#Посчитаем-количество-обзоров-и-среднюю-оценку-для-каждой-книги" data-toc-modified-id="Посчитаем-количество-обзоров-и-среднюю-оценку-для-каждой-книги-3.2"><span class="toc-item-num">3.2&nbsp;&nbsp;</span>Посчитаем количество обзоров и среднюю оценку для каждой книги</a></span></li><li><span><a href="#Определим-издательство,-которое-выпустило-наибольшее-число-книг-толще-50-страниц-(таки-образом-мы-исключим-из-анализа-брошюры)" data-toc-modified-id="Определим-издательство,-которое-выпустило-наибольшее-число-книг-толще-50-страниц-(таки-образом-мы-исключим-из-анализа-брошюры)-3.3"><span class="toc-item-num">3.3&nbsp;&nbsp;</span>Определим издательство, которое выпустило наибольшее число книг толще 50 страниц (таки образом мы исключим из анализа брошюры)</a></span></li><li><span><a href="#Определим-автора-с-самой-высокой-средней-оценкой-книг,-учитываются-книги-с-50-и-больше-оценок" data-toc-modified-id="Определим-автора-с-самой-высокой-средней-оценкой-книг,-учитываются-книги-с-50-и-больше-оценок-3.4"><span class="toc-item-num">3.4&nbsp;&nbsp;</span>Определим автора с самой высокой средней оценкой книг, учитываются книги с 50 и больше оценок</a></span></li><li><span><a href="#Посчитаем-среднее-количество-обзоров-от-пользователей,-которые-поставили-больше-50-оценок" data-toc-modified-id="Посчитаем-среднее-количество-обзоров-от-пользователей,-которые-поставили-больше-50-оценок-3.5"><span class="toc-item-num">3.5&nbsp;&nbsp;</span>Посчитаем среднее количество обзоров от пользователей, которые поставили больше 50 оценок</a></span></li><li><span><a href="#Выводы" data-toc-modified-id="Выводы-3.6"><span class="toc-item-num">3.6&nbsp;&nbsp;</span>Выводы</a></span></li></ul></li></ul></div>

# Исследование базы данных книжного приложения

<a id='start'></a>
# Описание проекта

**Цель проекта:**  
- Исследовать базу данных приобретенного сервиса для чтения книг.


**Задания:**  
- Посчитать, сколько книг вышло после 1 января 2000 года;  
- Для каждой книги посчитать количество обзоров и среднюю оценку;  
- Определить издательство, которое выпустило наибольшее число книг толще 50 страниц — так вы исключите из анализа брошюры;  
- Определить автора с самой высокой средней оценкой книг — учитывайте только книги с 50 и более оценками;  
- Посчитать среднее количество обзоров от пользователей, которые поставили больше 50 оценок.  

## Описание данных
<a id='describe'></a>

**Таблица `books`**

Содержит данные о книгах:

- `book_id` — идентификатор книги;
- `author_id` — идентификатор автора;
- `title` — название книги;
- `num_pages` — количество страниц;
- `publication_date` — дата публикации книги;
- `publisher_id` — идентификатор издателя.

**Таблица `authors`**

Содержит данные об авторах:

- `author_id` — идентификатор автора;
- `author` — имя автора.

**Таблица `publishers`**

Содержит данные об издательствах:

- `publisher_id` — идентификатор издательства;
- `publisher` — название издательства;

**Таблица `ratings`**

Содержит данные о пользовательских оценках книг:

- `rating_id` — идентификатор оценки;
- `book_id` — идентификатор книги;
- `username` — имя пользователя, оставившего оценку;
- `rating` — оценка книги.

**Таблица `reviews`**

Содержит данные о пользовательских обзорах на книги:

- `review_id` — идентификатор обзора;
- `book_id` — идентификатор книги;
- `username` — имя пользователя, написавшего обзор;
- `text` — текст обзора.

# Ознакомление с данными

In [1]:
# импортируем библиотеки
import pandas as pd
from sqlalchemy import create_engine

In [2]:
# устанавливаем параметры
db_config = {'user': 'praktikum_student', # имя пользователя
'pwd': 'Sdf4$2;d-d30pp', # пароль
'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
'port': 6432, # порт подключения
'db': 'data-analyst-final-project-db'} # название базы данных
connection_string = 'postgresql://{}:{}@{}:{}/{}'.format(db_config['user'],
 db_config['pwd'],
 db_config['host'],
 db_config['port'],
 db_config['db'])
# сохраняем коннектор
engine = create_engine(connection_string, connect_args={'sslmode':'require'})

## Таблица book

In [3]:
query = '''
            SELECT * 
            FROM books
        '''
books = pd.io.sql.read_sql(query, con = engine)
books.sample(5)

book_id  author_id                                         title  \
635      636        561                                The Art of War   
625      626        603                                    The Aeneid   
630      631        479  The Amber Spyglass  (His Dark Materials  #3)   
252      253        421                              For One More Day   
165      166        185                          Crime and Punishment   

     num_pages publication_date  publisher_id  
635        273       2005-01-11           264  
625        442       1990-06-16           309  
630        467       2003-09-09           172  
252        197       2006-09-26           122  
165        671       2002-12-31           210

In [4]:
books.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 6 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   book_id           1000 non-null   int64 
 1   author_id         1000 non-null   int64 
 2   title             1000 non-null   object
 3   num_pages         1000 non-null   int64 
 4   publication_date  1000 non-null   object
 5   publisher_id      1000 non-null   int64 
dtypes: int64(4), object(2)
memory usage: 47.0+ KB


[Описание данных](#describe)  

Таблица books содержит записи о 1000 книгах, столбца соответствуют описанию.

## Таблица authors

In [5]:
query = '''
            SELECT * 
            FROM authors
        '''
authors = pd.io.sql.read_sql(query, con = engine)
authors.sample(5)

author_id                                             author
295        296       Johanna Spyri/Beverly Cleary/Angelo  Rinaldi
382        383  Marcus Aurelius/Martin Hammond/Albert Wittstoc...
146        147                                       E.M. Forster
427        428                         Neal Stephenson/Guy Abadia
183        184  Fyodor Dostoyevsky/Anna Brailovsky/Joseph Fran...

In [6]:
authors.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 636 entries, 0 to 635
Data columns (total 2 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   author_id  636 non-null    int64 
 1   author     636 non-null    object
dtypes: int64(1), object(1)
memory usage: 10.1+ KB


Таблицы authors содержит 636 записей с именем автора и его идентификатором.

## Таблица publisher

In [7]:
query = '''
            SELECT * 
            FROM publishers
        '''
publishers = pd.io.sql.read_sql(query, con = engine)
publishers.sample(5)

publisher_id                               publisher
53             54                Bloomsbury Methuen Drama
314           315                   Vintage International
316           317                       Virago Press (UK)
101           102                                     Eos
272           273  Simon & Schuster; Media Tie-In edition

In [8]:
publishers.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 340 entries, 0 to 339
Data columns (total 2 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   publisher_id  340 non-null    int64 
 1   publisher     340 non-null    object
dtypes: int64(1), object(1)
memory usage: 5.4+ KB


Таблица publisher содержит 340 записей с назавниями издателей и их id-номера.

## Таблица ratings

In [9]:
query = '''
            SELECT * 
            FROM ratings
        '''
ratings = pd.io.sql.read_sql(query, con = engine)
ratings.sample(5)

rating_id  book_id     username  rating
4952       4953      763    patrick05       3
4754       4755      734   kevinquinn       5
1565       1566      284  amymckenzie       3
1837       1838      301      rmiller       5
4446       4447      700    michele09       5

In [10]:
ratings.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6456 entries, 0 to 6455
Data columns (total 4 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   rating_id  6456 non-null   int64 
 1   book_id    6456 non-null   int64 
 2   username   6456 non-null   object
 3   rating     6456 non-null   int64 
dtypes: int64(3), object(1)
memory usage: 201.9+ KB


Таблица ratings содержит 6456 записей о пользовательских оценках книг, столбцы соответствуют описанию данных.  
[Описание данных](#describe) 

## Таблица reviews

In [11]:
query = '''
            SELECT * 
            FROM reviews
        '''
reviews = pd.io.sql.read_sql(query, con = engine)
reviews.sample(5)

review_id  book_id    username  \
1829       1828      662  brianwoods   
1800       1799      652   jessica49   
2152       2152      774     ewerner   
2501       2502      893  kevinquinn   
1253       1253      457    nicole53   

                                                   text  
1829  Without west science quite social child campai...  
1800  Project charge around executive. Week thought ...  
2152  Range effort yard continue issue kid yeah. Ass...  
2501  Central newspaper happen show fish training ra...  
1253                       Tax hit though remain order.

In [12]:
reviews.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2793 entries, 0 to 2792
Data columns (total 4 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   review_id  2793 non-null   int64 
 1   book_id    2793 non-null   int64 
 2   username   2793 non-null   object
 3   text       2793 non-null   object
dtypes: int64(2), object(2)
memory usage: 87.4+ KB


Таблица reviews содержит 2793 записей с пользовательскими обзорами книг. Таблица соответствуею описанию данных.  
[Описание данных](#describe) 

In [13]:
# какие ещё таблицы доступны в базе
display(pd.io.sql.read_sql('''

SELECT *
FROM pg_catalog.pg_tables
WHERE schemaname != 'pg_catalog' AND
      schemaname != 'information_schema';

''', con = engine))


# смотрим на типы столбцов в интересующих таблицах
display(pd.io.sql.read_sql('''
SELECT 
    table_name, 
    column_name, 
    data_type, 
    is_nullable
FROM INFORMATION_SCHEMA.COLUMNS 
WHERE table_name IN ('books', 'authors', 'publishers', 'ratings', 'reviews');
''', con = engine))

schemaname           tablename         tableowner tablespace  hasindexes  \
0      public              orders    praktikum_admin       None        True   
1      public              visits    praktikum_admin       None        True   
2      public  advertisment_costs    praktikum_admin       None        True   
3      public             authors    praktikum_admin       None        True   
4      public              second  praktikum_student       None       False   
5      public            second_b  praktikum_student       None       False   
6      public          publishers    praktikum_admin       None        True   
7      public              author  praktikum_student       None        True   
8      public             reviews    praktikum_admin       None        True   
9      public             ratings    praktikum_admin       None        True   
10     public               books    praktikum_admin       None        True   

    hasrules  hastriggers  rowsecurity  
0      False        False        False  
1      False        False        False  
2      False        False        False  
3      False         True        False  
4      False        False        False  
5      False        False        False  
6      False         True        False  
7      False        False        False  
8      False         True        False  
9      False         True        False  
10     False         True        False

table_name       column_name data_type is_nullable
0      authors         author_id   integer          NO
1      authors            author      text         YES
2   publishers      publisher_id   integer          NO
3   publishers         publisher      text         YES
4      reviews         review_id   integer          NO
5      reviews           book_id   integer         YES
6      reviews          username      text         YES
7      reviews              text      text         YES
8      ratings         rating_id   integer          NO
9      ratings           book_id   integer         YES
10     ratings          username      text         YES
11     ratings            rating   integer         YES
12       books           book_id   integer          NO
13       books         author_id   integer         YES
14       books             title      text         YES
15       books         num_pages   integer         YES
16       books  publication_date      date         YES
17       books      publisher_id   integer         YES

# Задания

<a id='one'></a>  
## Сколько книг вышло после 1 января 2000 года  


In [14]:
query = '''
            SELECT COUNT(DISTINCT book_id) 
            FROM books
            WHERE publication_date > '2000-01-01'
        '''
pd.io.sql.read_sql(query, con = engine)

count
0    819

После 1 января 2000 года было выпущено 819 книг.

<a id='two'></a>  
## Посчитаем количество обзоров и среднюю оценку для каждой книги  


In [15]:
query = '''
            SELECT b.book_id,
                   b.title,
                   rev.rev_count,
                   ROUND(rat.avg_rating, 2) AS avg_rating
            FROM books AS b
            
            LEFT JOIN
              (SELECT book_id,
                        COUNT(review_id) AS rev_count
                        FROM reviews
                        GROUP BY book_id) AS rev ON rev.book_id=b.book_id
            LEFT JOIN
              (SELECT book_id,
                       AVG(rating) AS avg_rating
                       FROM ratings
                       GROUP BY book_id) AS rat ON rat.book_id=b.book_id                      
                        
        
            
        '''
agg_books = pd.io.sql.read_sql(query, con = engine)
agg_books

book_id                                              title  rev_count  \
0        652          The Body in the Library (Miss Marple  #3)        2.0   
1        273                                          Galápagos        2.0   
2         51                           A Tree Grows in Brooklyn        5.0   
3        951  Undaunted Courage: The Pioneering First Missio...        2.0   
4        839                                        The Prophet        4.0   
..       ...                                                ...        ...   
995       64                                Alice in Wonderland        4.0   
996       55          A Woman of Substance (Emma Harte Saga #1)        2.0   
997      148                                          Christine        3.0   
998      790  The Magicians' Guild (Black Magician Trilogy  #1)        2.0   
999      828                           The Plot Against America        2.0   

     avg_rating  
0          4.50  
1          4.50  
2          4.25  
3          4.00  
4          4.29  
..          ...  
995        4.23  
996        5.00  
997        3.43  
998        3.50  
999        3.00  

[1000 rows x 4 columns]

In [16]:
agg_books[['rev_count','avg_rating']].mean()

rev_count     2.809859
avg_rating    3.899040
dtype: float64

В среднем на книгу приходится 3 обзора, а средний рейтинг всех книг составляет 3,9 балла.


<a id='three'></a>  
## Определим издательство, которое выпустило наибольшее число книг толще 50 страниц (таки образом мы исключим из анализа брошюры)  


In [17]:
query = '''
        SELECT p.publisher_id,
               p.publisher,
               COUNT(b.book_id) AS count_books
        FROM books AS b
        LEFT JOIN publishers AS p ON p.publisher_id=b.publisher_id
        WHERE b.num_pages>50
        GROUP BY p.publisher_id,
                 p.publisher
        ORDER BY count_books DESC
        LIMIT 1
        '''
pd.io.sql.read_sql(query, con = engine)

publisher_id      publisher  count_books
0           212  Penguin Books           42

Больше всего книг (42 книги) толще 50 страниц выпустило издательство Penguin Books.

<a id='four'></a>  
## Определим автора с самой высокой средней оценкой книг, учитываются книги с 50 и больше оценок


In [18]:
query = '''
        SELECT a.author_id,
               a.author,
               ROUND(AVG(r.rating), 2) AS avg_ratings
        FROM books AS b
        LEFT JOIN ratings AS r ON r.book_id=b.book_id
        LEFT JOIN authors AS a ON a.author_id=b.author_id
        WHERE r.book_id IN
            (SELECT book_id
             FROM ratings
             GROUP BY book_id
             HAVING COUNT(rating_id)>=50)
        GROUP BY a.author_id,
                 a.author
        ORDER BY avg_ratings DESC
        LIMIT 1
        '''
pd.io.sql.read_sql(query, con = engine)

author_id                      author  avg_ratings
0        236  J.K. Rowling/Mary GrandPré         4.29

Джоан Роулинг в соавторстве с иллюстраторшей Мэри Гранпрэ имеют самый высокий рейтинг для своих книг.

<a id='five'></a>
## Посчитаем среднее количество обзоров от пользователей, которые поставили больше 50 оценок  


In [19]:
query = '''
        
        SELECT ROUND(AVG(coun_rev)) as avg_reviews
        FROM (SELECT username,
                COUNT(review_id) as coun_rev
                FROM reviews
                GROUP BY username
                HAVING username in (SELECT username
                                    FROM ratings
                                    GROUP BY username
                                    HAVING COUNT(rating_id)>50
                                    )
            ) as counter
        '''
pd.io.sql.read_sql(query, con = engine)

avg_reviews
0         24.0

Активные пользователи, которые поставили больше 50 оценок, в среднем пишут по 24 обзора.

## Выводы

В нашей базе данных содержится 5 таблиц.  
**books** - содержит о писание 1000 книг;  
**authors** - содержит информацию об 636 авторах;  
**publishers** - содержит информацию о 340 издательствах;  
**ratings** - содержит информацию о 6456 пользовательских оценок;   
**reviews** - содержит информацию о 2793 пользовательских обзорах.  

Таблицы полностью соответствуют [описанию](#describe).  


В ходе работы [для каждой книги было посчитано количество обзоров и найдена средняя оценка](#two)  и были даны ответы на 4 вопроса:  

- [Сколько книг вышло после 1 января 2000 года?](#one)  

*819 книг было выпущено с 1 января 2000 года*  




- [Какое издательство выпустило наибольшее количество книг с более чем 50 оценками?](#three)  

*Это издательство Penguin Books, которое выпустило 42 книги из нашей базы данных*  


- [У какого автора самая высокая средняя оценка книг?](#four)  

*Джоан Роулинг в соавторстве с иллюстраторшей Мэри Гранпрэ со средней оценкой 4.29 балла лидируют по данному показателю*  

- [Какое среднее количество обзоров пишут пользователи, которые поставили больше 50 оценок?](#five)  

*В среднем такие пользователь пишут 24 обзора*

[В начало](#start)